lib/explorer/backend/lazy_series.ex

```diff
@@ -0,0 +1,113 @@
+defmodule Explorer.Backend.LazySeries do
+  @moduledoc """
+  This is an opaque implementation of a Series.
+
+  It represents an operation with its arguments.
+  """
+  alias Explorer.Series
+  alias Explorer.Backend
+
+  @behaviour Explorer.Backend.Series
+
+  defstruct op: nil, args: []
+
+  @operations [
+    column: 1,
+    eq: 2,
+    neq: 2,
+    gt: 2,
+    gt_eq: 2,
+    lt: 2,
+    lt_eq: 2,
+    binary_and: 2,
+    binary_or: 2
+  ]
+  @comparison_ops_with_numbers [:eq, :neq, :gt, :gt_eq, :lt, :lt_eq]
+
+  @doc false
+  def new(op, args) do
+    %__MODULE__{op: op, args: args}
+  end
+
+  @doc false
+  def operations, do: @operations
+
+  # Implements all the comparison operations that
+  # accepts Series or number on the right-hand side.
+  for op <- @comparison_ops_with_numbers do
+    @impl true
+    def unquote(op)(%Series{} = left, %Series{} = right), do: unquote(op)(left, right.data)
+
+    def unquote(op)(%Series{} = left, value) do
+      data = new(unquote(op), [left.data, value])
+
+      Backend.Series.new(data, :boolean)
+    end
```

Comment:
I wonder if we should make this explicit?

```elixir
    def unquote(op)(%Series{data: left}, %Series{data: right}) do
      data = new(unquote(op), [left, right])
      Backend.Series.new(data, :boolean)
    end

    def unquote(op)(%Series{data: data}, value) when is_number(value) do
      data = new(unquote(op), [data, value])
      Backend.Series.new(data, :boolean)
    end
```

Is there anything we accept besides numbers?